### PR TITLE
feat!: enable read cache and write cache when using remote object stores

### LINKED ISSUE #4813: Enable read/write cache by default if using remote object stores
### What type of enhancement is this?

Performance

### What does the enhancement do?

The write cache and object store read cache significantly affect the performance. We should enable these caches if the storage is remote object stores.

### Implementation challenges

There are several challenges:
- The default path of the cache.
- The default capacity of the cache. We may infer this from memory or disk. We may not always get the correct or proper disk size. Using memory size may be easier.
- The mito engine may need to deal with multiple storages.

### PR DESCRIPTION
**Pull request by killme2008**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Close #4813 

## What's changed and what's your intention?

* Enable read and write cache by default for remote object storages.  
* Set the default read cache path to `{data_home}/object_cache/read` and create subdirectories for multiplex storages by their names.  
* Set the default write cache path to `{data_home}/object_cache/write`, as the write cache key, `IndexKey`, includes `region_id`, eliminating the need for subdirectories for multiplex storages.
* Change the default cache capacity from 1Gib to 5Gib.

@evenyag Please take a look.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

## Release Notes

- **New Features**
	- Enhanced configuration documentation for object storage and caching.
	- Added new test functions for OpenTelemetry Protocol (OTLP) metrics, traces, and logs.
	- Introduced a new error variant for directory creation failures.

- **Improvements**
	- Default cache size increased from 1GiB to 5GiB for object storage.
	- Clarified and updated configuration paths for experimental write cache.
	- Improved error handling and logging for cache directory creation.

- **Bug Fixes**
	- Adjusted cache path handling for S3 storage type to prevent unintended caching.

### CODE REVIEW

**Comment by killme2008** [comment]
@coderabbitai review

File: src/datanode/src/config.rs [hunk 1]
```diff
@@ -60,6 +60,14 @@ impl ObjectStoreConfig {
         }
     }
 
+    /// Returns true when it's a remote object storage such as AWS s3 etc.
+    pub fn is_object_storage(&self) -> bool {
+        matches!(
+            self,
+            Self::S3(_) | Self::Oss(_) | Self::Azblob(_) | Self::Gcs(_)
+        )
+    }
```
Comment (waynexia) [inline src/datanode/src/config.rs:69]: Nit: how about only excluding `Self::File`? This is not likely to change compared to the OSS services.

**Review by waynexia — APPROVED** [review]
Others LGTM.

Is it necessary to change the default path? I'm ok with the new naming but since it's a breaking change

**Comment by killme2008** [comment]
> Others LGTM.
> 
> Is it necessary to change the default path? I'm ok with the new naming but since it's a breaking change

I was hesitant as well, but I believe that failing to standardize the paths for read and write caches will lead to growing chaos in `data_home`. It's wiser to implement these breaking changes sooner rather than later.

**Review by WenyXu — APPROVED** [review]

File: src/datanode/src/config.rs (hunk 1 above)
Comment (WenyXu) [reply src/datanode/src/config.rs:69]: +1

**Comment by killme2008** [comment]
@evenyag Please help to take a look

File: src/datanode/src/config.rs [hunk 2]
```diff
@@ -60,6 +60,14 @@ impl ObjectStoreConfig {
         }
     }
 
+    /// Returns true when it's a remote object storage such as AWS s3 etc.
+    pub fn is_object_storage(&self) -> bool {
+        matches!(
+            self,
+            Self::S3(_) | Self::Oss(_) | Self::Azblob(_) | Self::Gcs(_)
+        )
```
Comment (evenyag) [inline src/datanode/src/config.rs:68]:
```suggestion
    pub fn is_object_storage(&self) -> bool {
        !matches!(self, Self::File(_))
```

File: src/datanode/src/datanode.rs [hunk 3]
```diff
@@ -399,10 +399,17 @@ impl DatanodeBuilder {
         for engine in &opts.region_engine {
             match engine {
                 RegionEngineConfig::Mito(config) => {
+                    let mut config = config.clone();
+                    if opts.storage.is_object_storage() {
+                        // Enable the write cache when setting object storage
+                        config.enable_experimental_write_cache = true;
+                        info!("Configured 'enable_experimental_write_cache=true' for mito engine.");
+                    }
```
Comment (evenyag) [inline src/datanode/src/datanode.rs:407]: Can we do this inside `Self::build_mito_engine()`?

File: src/mito2/src/worker.rs [hunk 4]
```diff
@@ -373,6 +373,12 @@ async fn write_cache_from_config(
     // TODO(yingwen): Remove this and document the config once the write cache is ready.
     warn!("Write cache is an experimental feature");
 
+    tokio::fs::create_dir_all(Path::new(&config.experimental_write_cache_path))
+        .await
+        .context(CreateDirSnafu {
+            dir: &config.experimental_write_cache_path,
+        })?;
```
Comment (evenyag) [inline src/mito2/src/worker.rs:380]: nitpick: Only do this when `config.enable_experimental_write_cache` is true

Comment (killme2008) [reply src/mito2/src/worker.rs:380]: Yes, it's checked above.

File: src/mito2/src/config.rs [hunk 5]
```diff
@@ -233,7 +234,8 @@ impl MitoConfig {
 
         // Sets write cache path if it is empty.
         if self.experimental_write_cache_path.is_empty() {
```
Comment (evenyag) [inline src/mito2/src/config.rs:236]:
```suggestion
        if self.experimental_write_cache_path.trim().is_empty() {
```

**Pull request merged**